ophyd/controls/device.py

```
@@ -0,0 +1,318 @@
+
+    @property
+    def connected(self):
+        return all(signal.connected for name, signal in self._signals.items())
+
+    def read(self):
+        # map names ("data keys") to actual values
+        values = {}
+        for name in self.read_signals:
+            signal = getattr(self, name)
+            values.update(signal.read())
+
+        return values
+
+    def describe(self):
+        desc = {}
+        for name in self.read_signals:
+            signal = getattr(self, name)
+            desc.update(signal.describe())
+
+        return desc
+
+    def stop(self):
+        "to be defined by subclass"
+        pass
+
+    def trigger(self):
+        "to be defined by subclass"
+        pass
+
+
+class OphydDevice(DeviceBase, OphydObject):
+    SUB_ACQ_DONE = 'acq_done'  # requested acquire
+
+    def __init__(self, prefix=None, read_signals=None,
+                 name=None, alias=None):
+        if name is None:
+            name = prefix
+
+        OphydObject.__init__(self, name=name, alias=alias)
+        DeviceBase.__init__(self, prefix, read_signals=read_signals)
+
+        # set should work using signature-stuff
+
+    @property
+    def trigger_signals(self):
+        names = [attr for cpt, attr in self._sig_attrs.items()
+                 if cpt.trigger_value is not None]
+
+        return [getattr(self, name) for name in names]
+
+    def trigger(self, **kwargs):
+        """Start acquisition"""
+        # TODO mass confusion here
+        signals = self.trigger_signals
+        if len(signals) > 1:
+            raise NotImplementedError('TODO more than one trigger')
+        elif len(signals) == 0:
+            raise RuntimeError('Device has no trigger signal(s)')
+
+        acq_signal = signals[0]
+        status = DeviceStatus(self)
+        self.subscribe(status._finished,
+                       event_type=self.SUB_ACQ_DONE, run=False)
+
+        def done_acquisition(**kwargs):
```

This silently eats the kwargs. It should propagate them or, more likely, not accept them.